File: frontend/src/api/messages.ts
import { apiFetch } from './httpClient'

// [Task 28][SEND-*][MSGX-*] /messages 발송 만들기의 얇은 클라이언트 — 경로·형태만 안다.
// 백엔드 계약: backend/app/routers/messages.py
//   GET /messages · POST /messages · DELETE /messages/scheduled/{id}
// ⛔ 실제 배달·발송 결과·재시도·명단 열람은 Task 30 — 이 모듈은 만들기까지만 안다.

/** 화면 종류값(SEND-KIND). 저장값(transactional/marketing)은 서버가 매핑한다. */
export type MessageKind = 'transactional' | 'marketing'
/** 보내는 방법(SEND-CH-01). push_sms=앱+문자폴백(기본) / push=앱만 / sms=모두 문자. */
export type MessageChannel = 'push_sms' | 'push' | 'sms'

/** 예약해 둔 발송 한 줄(SEND-LIST-01 위 구역). */
export interface ScheduledRow {
  id: string
  kind: MessageKind
  body: string | null
  channel: string | null
  scheduled_at: string
  target_count: number | null
  status: string
}

/** 배치별 상태 집계(SEND-RESULT-05·12). 한 번의 발송(대상 N명)의 상태 넷을 센 것. */
export interface SendResultCounts {
  발송중: number
  도달: number
  재시도중: number
  실패: number
}

/** 보낸 발송 한 줄(SEND-LIST-06 = 한 배치). 발송 결과는 result 집계로 온다(Task 30). */
export interface SentRow {
  id: string
  kind: MessageKind
  body: string | null
  channel: string
  /** [SEND-RESULT-09] 사용자가 고른 원래 3값(push_sms/push/sms). */
  requested_channel: string | null
  sender_staff_id: string | null
  target_count: number | null
  sent_at: string
  /** [SEND-RESULT-12] 배치별 상태 건수. */
  result: SendResultCounts
}

/** [SEND-FAIL-07] 안 닿은 한 사람 — 이름·(마스킹된)번호·왜 안 갔나. */
export interface FailedItem {
  id: string
  /** [SEND-FAIL-08] '번호 고쳐야 함' 탭의 [환자 열기]가 /patients/:id 로 갈 목적지. */
  patient_id: string | null
  name: string
  phone: string
  failure_code: string | null
  notification_type: string
  /** [SEND-FAIL-09] 지난 발송에서 이미 실패한 번호(접어둔다). */
  already_known: boolean
}

/** [SEND-FAIL-02] 안 닿은 명단 두 무리 — 지금 전화 / 번호 고쳐야 함. */
export interface FailedList {
  call_now: FailedItem[]
  fix_number: FailedItem[]
}

export interface SentPage {
  rows: SentRow[]
  has_more: boolean
  next_cursor: string | null
  order: string[]
}

/** GET /messages — 예약/보낸 두 구역 + 자동 발송 건수(SEND-LIST-01·02·08). */
export interface MessagesView {
  scheduled: ScheduledRow[]
  sent: SentPage
  auto_count: number
}

/** POST /messages 응답(EnqueueResult). 야간 광고 차단이면 night_blocked+suggested_at. */
export interface EnqueueResult {
  target_count: number
  sms_count: number | null
  marketing_excluded: number
  notification_ids: string[] | null
  scheduled_id: string | null
  night_blocked: boolean
  suggested_at: string | null
}

export type RecipientsSpec = { patient_ids: string[] } | { all: true }

export interface SendInput {
  kind: MessageKind
  recipients_spec: RecipientsSpec
  channel: MessageChannel
  body: string
  /** [SEND-LATER-01][MSGX-SCHED-01] 예약 발송 시각(ISO8601, KST). 없으면 즉시. */
  scheduled_at?: string | null
}

export function getMessages(cursor?: string | null): Promise<MessagesView> {
  const qs = cursor ? `?cursor=${encodeURIComponent(cursor)}` : ''
  return apiFetch<MessagesView>(`/messages${qs}`)
}

export function sendMessage(input: SendInput): Promise<EnqueueResult> {
  return apiFetch<EnqueueResult>('/messages', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
}

/** [MSGX-SCHED-02] pending 예약만 취소. 서버가 이미 종료된 예약을 409로 막는다. */
export function cancelScheduled(scheduledId: string): Promise<{ status: string }> {
  return apiFetch<{ status: string }>(`/messages/scheduled/${scheduledId}`, {
    method: 'DELETE',
  })
}

/** [SEND-FAIL-01] 한 배치의 안 닿은 명단(두 무리). batchId = 보낸 줄의 id. */
export function getFailedList(batchId: string): Promise<FailedList> {
  return apiFetch<FailedList>(`/messages/${batchId}/failed`)
}

/** [SEND-BADGE-01] 사이드바 숫자 — 전화해야 할 미처리 실패 건수. */
export function getBadgeCount(): Promise<{ count: number }> {
  return apiFetch<{ count: number }>('/messages/badge-count')
}

/** [SEND-BADGE-06] 처리 표시로 배지를 줄인다. */
export function markHandled(notificationId: string): Promise<{ status: string }> {
  return apiFetch<{ status: string }>(`/messages/${notificationId}/mark-handled`, {
    method: 'POST',
  })
}

File: backend/app/routers/messages.py
"""[Task 28][SEND-*][MSGX-*] /messages 발송 만들기 라우터 — 접수직원·관리자만.

⚠️ 코디 배선 필요: main.py에 `app.include_router(messages.router)`를 등록해야 노출된다.
   이 태스크는 만들기(enqueue)·목록·예약 취소까지다 — 실제 배달·결과·재시도는 Task 30.
"""
import hmac
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.core.config import settings
from app.core.security import StaffContext, require_role
from app.services import message_service

router = APIRouter(tags=["messages"])

_ROLES = ("receptionist", "admin")  # SEND-DOOR-07 — 의사 없음


class SendIn(BaseModel):
    kind: str
    recipients_spec: dict
    channel: str
    body: str
    scheduled_at: datetime | None = None


class SolapiReport(BaseModel):
    """[SEND-RESULT-02] SOLAPI 웹훅 리포트 한 건. 웹훅 본문은 이 객체의 배열이다.

    SOLAPI 실제 필드명 그대로: messageId(발송 시 저장한 provider_message_id), statusCode
    (`"4000"`=수신완료=도달, 그 외 종결 코드=실패). 나머지 필드(groupId·dateReported 등)는
    받되 쓰지 않는다(추가 필드 무시).
    """
    model_config = {"extra": "ignore"}
    messageId: str
    statusCode: str
    statusMessage: str | None = None


# [SEND-RESULT-02] SOLAPI 리포트 성공 코드 — 수신완료.
_SOLAPI_DELIVERED_CODE = "4000"


def _page_dto(page) -> dict:
    return {"rows": page.rows, "has_more": page.has_more,
            "next_cursor": page.next_cursor, "order": list(page.order)}


def _result_dto(res) -> dict:
    return {"target_count": res.target_count, "sms_count": res.sms_count,
            "marketing_excluded": res.marketing_excluded,
            "notification_ids": res.notification_ids, "scheduled_id": res.scheduled_id,
            "night_blocked": res.night_blocked, "suggested_at": res.suggested_at}


@router.post("/messages/status-callback")
async def status_callback(
    reports: list[SolapiReport],
    token: str | None = None,
) -> dict:
    """[SEND-RESULT-02][보안 F-03] SOLAPI 웹훅 수신 — 리포트 배열 + URL 토큰 인증.

    SOLAPI 웹훅은 커스텀 헤더가 아니라 **등록 URL에 심은 토큰**(`?token=`)으로 인증한다.
    제공자만 아는 토큰을 상수시간 비교해 위조 콜백을 막는다(시크릿 미설정 시 fail-closed).
    검증 실패·모르는 콜백 모두 같은 응답({"status":"ok"})을 돌려준다(ID oracle 제거).

    본문은 리포트 객체 **배열**이며, 각 건의 statusCode를 도달/실패로 매핑해 messageId
    (=발송 시 저장한 provider_message_id)로 줄을 찾아 상태를 굴린다.
    ⚠️ 4000 외 코드의 정확한 성공/실패 구분은 실발송 1건으로 최종 확정 대상(보수적으로 실패 처리).
    """
    secret = settings.solapi_webhook_secret
    if not secret or token is None or not hmac.compare_digest(token, secret):
        return {"status": "ok"}
    for r in reports:
        if r.statusCode == _SOLAPI_DELIVERED_CODE:
            await message_service.handle_status_callback(
                provider_message_id=r.messageId, status="delivered")
        else:
            await message_service.handle_status_callback(
                provider_message_id=r.messageId, status="failed", failure_code=r.statusCode)
    return {"status": "ok"}


@router.get("/messages/badge-count")
async def badge_count(
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[SEND-BADGE-01] 사이드바 숫자 — 전화해야 할 미처리 실패 건수."""
    return {"count": await message_service.badge_count(staff)}


@router.get("/messages/{batch_id}/failed")
async def failed_list(
    batch_id: UUID,
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[SEND-FAIL-02·06·07] 안 닿은 명단 — '지금 전화'·'번호 고쳐야 함' 두 무리."""
    return await message_service.failed_list(staff, batch_id)


@router.post("/messages/{notification_id}/mark-handled")
async def mark_handled(
    notification_id: UUID,
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[SEND-BADGE-06] 처리 표시로 배지를 줄인다."""
    return await message_service.mark_handled(staff, notification_id)


@router.get("/messages")
async def list_messages(
    cursor: str | None = None,
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[SEND-DOOR-02][SEND-LIST-01·08] 예약해 둔 것·보낸 것 두 구역 + 자동 발송 건수."""
    out = await message_service.list_messages(staff, cursor=cursor)
    return {"scheduled": out["scheduled"], "sent": _page_dto(out["sent"]),
            "auto_count": out["auto_count"]}


@router.post("/messages")
async def send_message(
    body: SendIn,
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[SEND-BOX-*][SEND-NIGHT-02] 발송을 큐에 넣는다. 야간 광고 차단이면 200 + 제안 시각."""
    res = await message_service.enqueue_send(
        staff, kind=body.kind, recipients_spec=body.recipients_spec,
        channel=body.channel, body=body.body, scheduled_at=body.scheduled_at)
    return _result_dto(res)


@router.delete("/messages/scheduled/{scheduled_id}")
async def cancel_scheduled(
    scheduled_id: UUID,
    staff: StaffContext = Depends(require_role(*_ROLES)),
) -> dict:
    """[MSGX-SCHED-02][SEND-LATER-05] pending 예약만 취소, 취소자·시각 기록."""
    return await message_service.cancel_scheduled(staff, scheduled_id, expected_status="pending")
